File: BackUpForABook/UIControls/TRClasses/TRSliceTitle.m
//
//  TRSliceTitle.m
//  ATreeRingMap
//
//  Created by Hui Wang on 7/16/14.
//  Copyright (c) 2014 AnnieLyticx. All rights reserved.
//

#import "TRSliceTitle.h"
#import <CoreText/CoreText.h>

@interface TRSliceTitle (PrivateMethods)

// paint slice title
-(void)paintSliceTitle:(CGContextRef)ctx
                 title:(NSMutableString*)title
                radius:(float)radius
                  left:(float)left
                 right:(float)right
                center:(CGPoint)origin
           frameHeight:(int)height;

@end

@implementation TRSliceTitle

// private data

// init
-(id)init
{
    self = [super init];
    
    textPainter = [[ANCircleText alloc] init];
    
    _title = nil;
    
    _size = 18;
    _textAttributes = [NSMutableDictionary dictionary];
    _style = RingTextStyle_Ori_Left2Right | RingTextStyle_AlignMiddle;
    
    // set inital attributes
    CTFontRef lbFont = CTFontCreateWithName(CFSTR("Helvetica-Bold"), 10.0, NULL);
    
    _textAttributes[NSFontAttributeName] = (__bridge id)lbFont;
    CFRelease(lbFont);

    // text area color (fill area)
    _textAttributes[NSForegroundColorAttributeName] = (id)[[UIColor colorWithRed:.0 green:.0 blue: .0 alpha:.4] CGColor];

    // text edge color (stroking area)
    _textAttributes[NSStrokeColorAttributeName] = (id)[[UIColor colorWithRed:.0 green:.8 blue:.0 alpha:0.4] CGColor];

    // nagativr value both stroke and fill the text.  Positive value for "hollow" presentation only stroke the edges
    _textAttributes[NSStrokeWidthAttributeName] = (id)[NSNumber numberWithFloat:-3.0];
    // end of attributed string properties

    // run time data
    _pointsPerFontSize = 1.0;
    
    return self;
}

// method to setup string attributes
-(void)setStringAttributes:(NSString*)font
                      size:(float)size
           foregroundColor:(UIColor*)foregroundColor
               strokeColor:(UIColor*)strokeColor
               strokeWidth:(float)strokeWidth
{
    // font
    CTFontRef lbFont = CTFontCreateWithName((CFStringRef)font, size, NULL);
    
    _textAttributes[NSFontAttributeName] = (__bridge id)lbFont;
    CFRelease(lbFont);
    
    // text area color (fill area)
    _textAttributes[NSForegroundColorAttributeName] = (id)[foregroundColor CGColor];
    
    // text edge color (stroking area)
    _textAttributes[NSStrokeColorAttributeName] = (id)[strokeColor CGColor];
    
    // nagativr value both stroke and fill the text.  Positive value for "hollow" presentation only stroke the edges
    _textAttributes[NSStrokeWidthAttributeName] = (id)[NSNumber numberWithFloat:strokeWidth];
    // end of attributed string properties
}

-(void)paint:(CGContextRef)ctx
      radius:(float)radius
        left:(float)left
       right:(float)right
      center:(CGPoint)origin
 frameHeight:(int)height
{
    if (_title != nil)
    {
        // draw title string in the center of given range
        [self paintSliceTitle:ctx title:_title radius:radius left:left right:right center:origin frameHeight:height];
    }
}

// paint title at given position
-(void)paintSliceTitle:(CGContextRef)ctx
                 title:(NSMutableString*)title
                radius:(float)radius
                  left:(float)left
                 right:(float)right
                center:(CGPoint)origin
           frameHeight:(int)height
{
    // make attributed string for the title first
    NSMutableAttributedString *attString = [[NSMutableAttributedString alloc] initWithString:title attributes:_textAttributes];
    
    [textPainter paintCircleText:ctx text:attString style:_style radius:radius width: _size * _pointsPerFontSize left:left right:right center:origin];
    
    return ;
}

// hit test.  Potential hit objects: slider, begin arrow and end arrow
-(HitObj)HitTest:(CGPoint)atPoint
          radius:(float)radius
          center:(CGPoint)origin
{
    HitObj hitObj;
    
    hitObj.hitObject = TRObjs_None;
    
    return hitObj;
}

@end

File: BackUpForABook/UIControls/TRClasses_swift/TRSliceTitle.swift
//
//  TRSliceTitle.swift
//  TreeRingMap_Swift
//
//  Created by dingf on 17/3/16.
//  Copyright © 2017年 MH.dingf. All rights reserved.
//

import UIKit

class TRSliceTitle: NSObject {
    var textPainter: ANCircleText?
    var size: Float?
    var style: RingTextStyle?
    var textAttributes = [String : Any]()
    var pointsPerFontSize: Float?
    var title: String?
    
    override init() {
        super.init()
        textPainter = ANCircleText.init()
        size = 18
        style = .alignMiddle
        textAttributes[NSFontAttributeName] = CTFontCreateWithName("Helvetica-Bold" as CFString, 10.0, nil)
        textAttributes[NSForegroundColorAttributeName] = UIColor(red: 0.0, green: 0.0, blue: 0.0, alpha: 0.4)
        textAttributes[NSStrokeColorAttributeName] = UIColor(red: 0.0, green: 0.8, blue: 0.0, alpha: 0.4)
        textAttributes[NSStrokeWidthAttributeName] = NSNumber(value: -3.0)
        pointsPerFontSize = 1.0
        
    }
    
    func setStringAttributes(_ font: String, size: Float, foregroundColor: UIColor, strokeColor: UIColor, strokeWidth: Float){
        let lbFont = CTFontCreateWithName(font as CFString, CGFloat(size), nil)
        textAttributes[NSFontAttributeName] = lbFont
        textAttributes[NSForegroundColorAttributeName] = foregroundColor.cgColor
        textAttributes[NSStrokeColorAttributeName] = strokeColor.cgColor
        textAttributes[NSStrokeWidthAttributeName] = NSNumber(value: strokeWidth)
        
    }
    
    // paint title at given position
    func paintSliceTitle(_ ctx: CGContext, radius: Float, left: Float, right: Float, center: CGPoint,height: Int){
        // make attributed string for the title 
        let attString = NSMutableAttributedString.init(string: title!, attributes: textAttributes)
        textPainter?.paintCircleText(ctx, text: attString, style: style!, radius: CGFloat(radius), width: CGFloat(size! * pointsPerFontSize!), left: CGFloat(left), right: CGFloat(right), center: center)
    }

}
